drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,514 @@
+	}
+
+	kscan_npcx_drive_column(dev, KEYBOARD_COLUMN_DRIVE_NONE);
+
+	return key_event != 0U ? true : false;
+}
+
+static bool check_key_events(const struct device *dev)
+{
+	uint8_t matrix_new_state[CONFIG_KSCAN_NPCX_COLUMN_SIZE] = {0U};
+	struct kscan_npcx_data *const data = dev->data;
+	uint32_t cycles_now = k_cycle_get_32();
+	bool key_pressed = false;
+	uint8_t row_changed = 0U;
+	uint8_t deb_col;
+
+	if (++data->scan_cycles_idx >= SCAN_OCURRENCES) {
+		data->scan_cycles_idx = 0U;
+	}
+
+	data->scan_clk_cycle[data->scan_cycles_idx] = cycles_now;
+
+	/* Scan the matrix */
+	key_pressed = read_keyboard_matrix(dev, matrix_new_state);
+
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		LOG_DBG("U%x, P%x, N%x", data->matrix_unstable_state[c],
+			data->matrix_previous_state[c], matrix_new_state[c]);
+	}
+
+	/* Abort if ghosting is detected */
+	if (is_matrix_ghosting(matrix_new_state)) {
+		return key_pressed;
+	}
+
+	/*
+	 * The intent of this loop is to gather information related to key
+	 * changes.
+	 */
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		/* Check if there was an update from the previous scan */
+		row_changed = matrix_new_state[c] ^ data->matrix_previous_state[c];
+
+		if (!row_changed) {
+			continue;
+		}
+
+		for (int r = 0; r < CONFIG_KSCAN_NPCX_ROW_SIZE; r++) {
+			/*
+			 * Index all they keys that changed for each row
+			 * in order to debounce each key in terms of it
+			 */
+			if (row_changed & BIT(r)) {
+				data->scan_cycle_idx[c][r] = data->scan_cycles_idx;
+			}
+		}
+
+		data->matrix_unstable_state[c] |= row_changed;
+		data->matrix_previous_state[c] = matrix_new_state[c];
+	}
+
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		deb_col = data->matrix_unstable_state[c];
+
+		if (!deb_col) {
+			continue;
+		}
+
+		/* Debouncing for each row key occurs here */
+		for (int r = 0; r < CONFIG_KSCAN_NPCX_ROW_SIZE; r++) {
+			uint8_t mask = BIT(r);
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask)) {
+				continue;
+			}
+
+			/* Convert the clock cycle differences to usec */
+			uint32_t debt = k_cyc_to_us_floor32(
+				cycles_now - data->scan_clk_cycle[data->scan_cycle_idx[c][r]]);
+
+			/* Does the key requires more time to be debounced? */
+			if (debt < (row_bit ? data->deb_time_press : data->deb_time_rel)) {
+				/* Need more time to debounce */
+				continue;
+			}
+
+			data->matrix_unstable_state[c] &= ~row_bit;
+
+			/* Check if there was a change in the stable state */
+			if ((data->matrix_stable_state[c] & mask) == row_bit) {
+				/* Key state did not change */
+				continue;
+			}
+
+			/*
+			 * The current row has been debounced, therefore update
+			 * the stable state. Then, proceed to notify the
+			 * application about the keys pressed.
+			 */
+			data->matrix_stable_state[c] ^= mask;
+			if (data->callback) {
+				data->callback(dev, r, c, row_bit ? true : false);
+			}
+		}
+	}
```

Comment:
Maybe this block could go in its own function? Thinking `update_matrix_state` or something.

Reply:
Done.